[flang] handle alloca outside of entry blocks in MemoryAllocation

jeanPerier opened the pull request:
This patch  generalizes the MemoryAllocation pass (alloca -> heap) to handle fir.alloca regardless of their postion in the IR. Currently, it only dealt with fir.alloca in function entry blocks. The logic is placed in a utility that can be used to replace alloca in an operation on demand to whatever kind of allocation the utility user wants via callbacks (allocmem, or custom runtime calls to instrument the code...).

To do so, a concept of ownership, that was already implied a bit and used in passes like stack-reclaim, is formalized. Any operation with the LoopLikeInterface, AutomaticAllocationScope, or IsolatedFromAbove owns the alloca directly nested inside its regions, and they must not be used after the operation.

The pass then looks for the exit points of region with such interface, and use that to insert deallocation. If dominance is not proved, the pass fallbacks to storing the new address into a C pointer variable created in the entry of the owning region which allows inserting deallocation as needed, included near the alloca itself to avoid leaks when the alloca is executed multiple times due to block CFGs loops.

This should fix https://github.com/llvm/llvm-project/issues/88344.

In a next step, I will try to refactor lowering a bit to introduce lifetime operation for alloca so that the deallocation points can be inserted as soon as possible.

jeanPerier committed:
[flang] handle alloca outside of entry blocks in MemoryAllocation

tblah commented on flang/include/flang/Optimizer/Transforms/MemoryUtils.h:
```diff
@@ -0,0 +1,63 @@
+/// The callback should not delete the alloca, the utility will do it.
+using AllocaRewriterCallBack =
+    llvm::function_ref<mlir::Value(mlir::OpBuilder &, fir::AllocaOp,
+                                   bool /*allocaDominatesDeallocLocations*/)>;
```
nit: why is this commented?

jeanPerier replied:
I thought I could not name argument here and did not even try, thanks!

tblah commented on flang/include/flang/Optimizer/Transforms/MemoryUtils.h:
```diff
@@ -0,0 +1,63 @@
+    llvm::function_ref<void(mlir::Location, mlir::OpBuilder &, mlir::Value)>;
+
+/// Utility to replace fir.alloca by dynamic allocations inside \p parentOp.
+/// \p MustRewriteCallBack let the user control which fir.alloca should be
+/// replaced. \p AllocaRewriterCallBack let the user define how the new memory
+/// should be allocated. \p DeallocCallBack let the user decide how the memory
+/// should be deallocated. The boolean result indicate if the utility succeeded
+/// to replace all fir.alloca as requested by the user. Causes of failures are
+/// the presence of unregistered operations, or OpenMP/ACC recipe operation that
+/// returns memory allocated inside their region.
```
nit
```suggestion
/// \p MustRewriteCallBack lets the user control which fir.alloca should be
/// replaced. \p AllocaRewriterCallBack lets the user define how the new memory
/// should be allocated. \p DeallocCallBack lets the user decide how the memory
/// should be deallocated. The boolean result indicates if the utility succeeded
/// to replace all fir.alloca as requested by the user. Causes of failures are
/// the presence of unregistered operations, or OpenMP/ACC recipe operations that
/// return memory allocated inside their region.
```

tblah commented on flang/lib/Optimizer/Transforms/MemoryUtils.cpp:
```diff
@@ -0,0 +1,220 @@
+        return nullptr;
+      deallocationPoints.push_back(terminator);
+    }
+  // If the owningRegion did not adhere to the ReturnLike interface for its
+  // terminators, bail and do not attempt to translate it (we could maybe
+  // fallback to consider terminators with no block successor, but since all
+  // FIR, OpenACC, OpenMP, CUF, SCF operations with IsIsolatedFromAbove,
+  // AutomaticAllocationScope, or LoopLikeOpInterface have such terminators,
+  // avoid any untested complexity for now).
```
This seems to conflict with the comment in `isRegionTerminator`

jeanPerier replied:
It does, it is a left over, thanks for the catch.

tblah commented on flang/lib/Optimizer/Transforms/MemoryUtils.cpp:
```diff
@@ -0,0 +1,220 @@
+  if (deallocationPoints.empty())
+    return nullptr;
+
+  // Step 3: detect loops between the alloc and deallocation points.
+  // If such loop exists, the easy solution is to consider the alloc
+  // as a deallocation point of any previous allocation. This works
+  // because the alloc does not properly dominates itself, so the
+  // inserted deallocation will be conditional.
+  // For now, always assume there may always be a loop if any of the
+  // deallocation point does not dominate the alloca. It is
+  // conservative approach. Bringing lifetime markers above will reduce
+  // the false positive for alloca made inside if like constructs or CFG.
+  if (!allocDominatesDealloc(alloca, deallocationPoints))
+    deallocationPoints.push_back(alloca.getOperation());
+  return owningRegion;
+}
```
I'm having trouble understanding this bit. Please could you give a pseudo-IR example?

jeanPerier replied:
There are two things that are being described here: how block loops are detected, and how they are dealt with.
 
1. If the alloca is inside a block based loop, and the deallocation points are after the loop, the deallocation points will not be dominated by the alloca. So we can use dominance to detect block based loops with some false positive (alloca in "if-like" block CFG will also create cases where deallocation points are not dominated by the alloca).

2. if the alloca is inside a block based loop, and the deallocation points are after the loop, an extra deallocation must be added on the back edge to avoid memory leaks. The easiest is to insert a conditional deallocation right before the alloca: If the alloca has ever been executed before, the C pointer variable will be set and is deallocated right before executing the alloca again. This does not require any analysis to find the back edge and can be done in the "false positive" cases with no correctness consequences (in fact, it could always be done).

Here is an illustration:
```
subroutine test(n)
  integer :: n
  real :: res
  interface
    function create_array(i)
      real :: create_array(i)
    end function
  end interface
  res = 0
  do i=1,100
    res = res + sum(create_array(i))
    if (i>n) goto 1
  end do
1 call do_something(res)
end subroutine
```

Because of the goto, the do loop is lowered to block CFG (blocks + `cf.cond_br`). The alloca is contained inside a block for the loop body. The alloca owner is the func.func, since there are no region based operation for the loop. So the deallocation point that was found so far is the func.return. The func.return is not dominated by the block containing the alloca due to the cyclic aspect. Hence, a conditional deallocation will be inserted before the new allocmem.

```
  func.func @_QPtest(%arg0: !fir.ref<i32> {fir.bindc_name = "n"}) {
    cf.br ^bb1

  ^bb1:  // 2 preds: ^bb0, ^bb4
    // ... loop entry
    cf.cond_br %8, ^bb2, ^bb5
  ^bb2:  // pred: ^bb1
    %17 = fir.alloca !fir.array<?xf32>, %15 {bindc_name = ".result"}
    // ... SUM
    // ... cf.cond_br for the IF () GOTO 1
    cf.cond_br %38, ^bb3, ^bb4
  ^bb3:  // pred: ^bb2
    cf.br ^bb5
  ^bb4:  // pred: ^bb2
    // ... index increment and cf.br back to the loop entry
    cf.br ^bb1
  ^bb5:  // 2 preds: ^bb1, ^bb3
    fir.call @_QPdo_something(%6) fastmath<contract> : (!fir.ref<f32>) -> ()
    return
  }
```

Will be turned into:


```
  func.func @_QPtest(%arg0: !fir.ref<i32> {fir.bindc_name = "n"}) {
    // Init pointer variable to nullptr
    %0 = fir.alloca !fir.heap<!fir.array<?xf32>>
    %1 = fir.zero_bits !fir.heap<!fir.array<?xf32>>
    fir.store %1 to %0 : !fir.ref<!fir.heap<!fir.array<?xf32>>>
    cf.br ^bb1
  ^bb1:  // 2 preds: ^bb0, ^bb4
    // loop entry
    cf.cond_br %8, ^bb2, ^bb5
  ^bb2:  // pred: ^bb1
    // Conditionally dealloc the memory if pointer variable is set.
    %19 = fir.load %0 : !fir.ref<!fir.heap<!fir.array<?xf32>>>
    %20 = fir.convert %19 : (!fir.heap<!fir.array<?xf32>>) -> i64
    %21 = arith.cmpi ne, %20, %c0_i64 : i64
    fir.if %21 {
      fir.freemem %19 : !fir.heap<!fir.array<?xf32>>
    }
   // Allocate memory and set pointer variable to it.
    %22 = fir.allocmem !fir.array<?xf32>, %17 {bindc_name = ".result", uniq_name = ""}
    %23 = fir.convert %22 : (!fir.heap<!fir.array<?xf32>>) -> !fir.ref<!fir.array<?xf32>>
    fir.store %22 to %0 : !fir.ref<!fir.heap<!fir.array<?xf32>>>
    //  ...
    cf.cond_br %38, ^bb3, ^bb4
  ^bb3:  // pred: ^bb2
    cf.br ^bb5
  ^bb4:  // pred: ^bb2
    cf.br ^bb1
  ^bb5:  // 2 preds: ^bb1, ^bb3
    fir.call @_QPdo_something(%6) fastmath<contract> : (!fir.ref<f32>) -> ()
    // Conditionally dealloc the memory if pointer variable is set
    // ...
    fir.if %51 {
      fir.freemem %49 : !fir.heap<!fir.array<?xf32>>
    }
    return
  }
```

I rephrased the comment to be a bit more clear.

tblah replied:
Thank you for explaining. LGTM

tblah commented on flang/lib/Optimizer/Transforms/MemoryUtils.cpp:
```diff
@@ -0,0 +1,220 @@
+static mlir::Value castIfNeeed(mlir::Location loc, mlir::RewriterBase &rewriter,
+                               mlir::Type newType, mlir::Value value) {
+  if (value.getType() != newType)
+    return rewriter.create<fir::ConvertOp>(loc, newType, value);
```
Doesn't `fir::ConvertOp::fold` do this already?

jeanPerier replied:
It does, it is just so cheap to check and not needed most of the time that I hate polluting the IR with them (makes IR reading harder). I will use a shared utility to avoid doing that everywhere.

jeanPerier replied:
I moved/shared the logic with FIROpBuilder::createConvert to avoid the proliferation of this helper.

tblah commented on flang/lib/Optimizer/Transforms/MemoryUtils.cpp:
```diff
@@ -0,0 +1,220 @@
+  rewriter.create<fir::StoreOp>(loc, nullPtr, ptrVar);
+  // TODO: introducing a pointer compare op in FIR would help
+  // generating less IR here.
+  mlir::Type intPtrTy = rewriter.getI64Type();
```
nit: is there a less platform specific type we could use for pointers? If the target information is unavailable maybe an index will always be pointer-sized?

jeanPerier replied:
Index is an integer with the ["machine word size"](https://mlir.llvm.org/docs/Rationale/Rationale/#integer-signedness-semantics), it is not guaranteed to be always pointer sized. So it is not much better. I rather keep getI64Type() because it is easier to grep for (less hit, index is used a lot).

LLVM IR just uses icmp with pointer operand without any explicit cast to integer types. We likely want to have some operation to do that to avoid bothering with the target dependency for null pointer comparisons.

jeanPerier replied:
I moved/shared that with  fir::FIROpbuilder::getIntPtrType.

tblah commented on flang/lib/Optimizer/Transforms/MemoryUtils.cpp:
```diff
@@ -0,0 +1,220 @@
+      findDeallocationPointsAndOwner(alloca, deallocationPoints);
+  if (!owningRegion)
+    return false;
+  // return false;
```
nit

tblah reviewed (commented):
Really clean implementation here. Nice work!

psteinfeld reviewed (commented):
All builds and tests correctly.  But this change is supposed to fix #88344.  But after building flang-new with these changes, the example program in that issue still gets a seg fault in the same place.

jeanPerier commented:
> All builds and tests correctly. But this change is supposed to fix #88344. But after building flang-new with these changes, the example program in that issue still gets a seg fault in the same place.

Did you use the `-mmlir -fdynamic-heap-array` switch?

#88344 test is expected to segfault with limited stacks given flang is doing stack allocation by default in many cases (the test needs a huge temp). The bug was that `-mmlir -fdynamic-heap-array`  (which place stack temps on the heap) did not fix the situation. It does for me now (well, I did not wait for the reproducer to terminate, it is a torture test that does 100 billion multiplications and 10 million dynamic allocations.... but without the `-fdynamic-heap-array` option it crashes after a few seconds only).

jeanPerier committed:
Update flang/include/flang/Optimizer/Transforms/MemoryUtils.h

Co-authored-by: Tom Eccles <t@freedommail.info>

jeanPerier committed:
address comments

- move conditional convert logic creation into factory helper
- move target dependent int pointer type creation into helper
- remove obsolete comments

jeanPerier committed:
format

jeanPerier committed:
rephrase step 3 comment

psteinfeld commented:
> Did you use the `-mmlir -fdynamic-heap-array` switch?
Oops.  I missed that.

But I've rebuilt with the latest sources and tested the original test, and all looks good.

psteinfeld reviewed (approved):
All builds and tests correctly and looks good.

razvanlupusoru reviewed (approved):
Excellent approach and sensible implementation. Looks great!

vzakhari commented on flang/lib/Optimizer/Transforms/MemoryUtils.cpp:
```diff
@@ -0,0 +1,217 @@
+  // points, and add a deallocation point on the back edge to avoid memory
+  // leaks.
+  // To avoid complex CFG analysis, the detection uses the fact an alloca
+  // inside a loop does not dominate its deallocation points after the loop. So
```
Is this always true for unstructured loops? For example:
```
subroutine test(x, n)
  interface
     function create_array(i)
       real :: create_array(i)
     end function create_array
  end interface
  real :: x(:)
  integer :: n, i
  i = 1
10 x(i) = sum(create_array(i))
  i = i + 1
  if (i.lt.n) goto 10
end subroutine test
```

The `alloca` still dominates the `return`. It seems that extending the `MemoryAllocationOpt` pass this way will cause memory leaks for this case.  It seems a conservative approach might be to check if the alloca's block is reachable from itself (via unstructured branching), and use the indirect deallocation when it is.

vzakhari replied:
Thank you for addressing the issue, Jean! The custom cycle detector looks good to me.

vzakhari reviewed (commented):
Thank you, Jean!  It looks great, though, I think there might be regressions for some unstructured CFG cases.

tblah reviewed (approved):
LGTM once Slava's comment is addressed. Thanks for the updates

jeanPerier committed:
handle block cycles properly
